# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/HelperThreadState.h
# Commit: 2bc5949902ac
# Full Hash: 2bc5949902acc042205fc7e85260bbbc37dbedc7
# Author: Matthew Gaudet <mgaudet@mozilla.com>
# Date: 2021-09-18 09:46:57
# Regressor Bug: 1688788
# File Overlap Count: 1
# Description:
#   Bug 1688788 - Convert ScriptPreloader to manage Stencils, not Scripts. r=kmag,tcampbell
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D113013
# ==============================================================================

diff -r e660e2c3c3d1 -r 2bc5949902ac js/src/vm/HelperThreadState.h
--- a/js/src/vm/HelperThreadState.h	Sat Sep 18 01:29:45 2021 +0000
+++ b/js/src/vm/HelperThreadState.h	Sat Sep 18 01:29:45 2021 +0000
@@ -15,11 +15,14 @@
 
 #include "mozilla/Attributes.h"
 #include "mozilla/EnumeratedArray.h"
+#include "mozilla/RefPtr.h"  // RefPtr
 #include "mozilla/TimeStamp.h"
+#include "mozilla/Vector.h"  // mozilla::Vector
 
 #include "ds/Fifo.h"
 #include "frontend/CompilationStencil.h"  // CompilationStencil, ExtensibleCompilationStencil, CompilationGCOutput
 #include "js/CompileOptions.h"
+#include "js/experimental/JSStencil.h"
 #include "js/HelperThreadAPI.h"
 #include "js/TypeDecls.h"
 #include "threading/ConditionVariable.h"
@@ -60,8 +63,8 @@
   // The output is JSScript.
   ScriptDecode,
 
-  // The output is an array of JSScript.
-  MultiScriptsDecode,
+  // The output is an array of CompilationStencil.
+  MultiStencilsDecode,
 };
 enum class StartEncoding { No, Yes };
 
@@ -391,7 +394,7 @@
   bool generateLCovSources(JSContext* cx, ParseTask* parseTask);
   bool finishMultiParseTask(JSContext* cx, ParseTaskKind kind,
                             JS::OffThreadToken* token,
-                            MutableHandle<ScriptVector> scripts);
+                            mozilla::Vector<RefPtr<JS::Stencil>>* stencils);
 
   void mergeParseTaskRealm(JSContext* cx, ParseTask* parseTask,
                            JS::Realm* dest);
@@ -409,8 +412,9 @@
   UniquePtr<frontend::CompilationStencil> finishCompileToStencilTask(
       JSContext* cx, JS::OffThreadToken* token);
   JSScript* finishScriptDecodeTask(JSContext* cx, JS::OffThreadToken* token);
-  bool finishMultiScriptsDecodeTask(JSContext* cx, JS::OffThreadToken* token,
-                                    MutableHandle<ScriptVector> scripts);
+  bool finishMultiStencilsDecodeTask(
+      JSContext* cx, JS::OffThreadToken* token,
+      mozilla::Vector<RefPtr<JS::Stencil>>* stencils);
   JSObject* finishModuleParseTask(JSContext* cx, JS::OffThreadToken* token);
 
   frontend::CompilationStencil* finishStencilParseTask(
@@ -511,6 +515,10 @@
   // ParseTask.
   GCVector<JSScript*, 1, SystemAllocPolicy> scripts;
 
+  // For the multi-decode stencil case, holds onto the set of stencils produced
+  // offthread
+  mozilla::Vector<RefPtr<JS::Stencil>> stencils;
+
   // Holds the ScriptSourceObjects generated for the script compilation.
   GCVector<ScriptSourceObject*, 1, SystemAllocPolicy> sourceObjects;
 
@@ -564,12 +572,12 @@
   void parse(JSContext* cx) override;
 };
 
-struct MultiScriptsDecodeTask : public ParseTask {
+struct MultiStencilsDecodeTask : public ParseTask {
   JS::TranscodeSources* sources;
 
-  MultiScriptsDecodeTask(JSContext* cx, JS::TranscodeSources& sources,
-                         JS::OffThreadCompileCallback callback,
-                         void* callbackData);
+  MultiStencilsDecodeTask(JSContext* cx, JS::TranscodeSources& sources,
+                          JS::OffThreadCompileCallback callback,
+                          void* callbackData);
   void parse(JSContext* cx) override;
 };
 